File: hard/as_quick_as_a_flash/as.quick.as.a.flash.py
# Reuben Thorpe (2016), CodeEval [As Quick As A Flash v1.0]
from sys import argv

def quick_sort_pivots(seq):
    # Returns the number of pivots enacted by the quick sort algorithm
    pivot_count = 0
    N = len(seq) - 1

    def swap(seq, x, y):
        seq[x], seq[y] = seq[y], seq[x]


    def q_sort(seq, begin, end):
        # Code Evals implimentation of the quick sort algorithm
        if begin < end:
            nonlocal pivot_count
            pivot_count += 1
            pivot = seq[begin]
            pivot_pos = begin
            left = begin+1
            right = end

            while left < right:
                while (left < right and seq[right] >= pivot):
                    right -= 1

                if seq[right] < pivot:
                    swap(seq, pivot_pos, right)
                    pivot_pos = right

                if left < right:
                    while (left < right and seq[left] <= pivot):
                        left += 1

                    if seq[left] > pivot:
                        swap(seq, pivot_pos, left)
                        pivot_pos = left

            q_sort(seq, begin, pivot_pos-1)
            q_sort(seq, pivot_pos+1, end)

    q_sort(seq, 0, N)
    return(pivot_count)


if __name__ == "__main__":
    for line in open(argv[1], "r"):
        line = [int(num) for num in line.split()]
        print(quick_sort_pivots(line))

File: hard/as_quick_as_a_flash/as.quick.as.a.flash.cpp
// Reuben Thorpe (2016), CodeEval [As Quick As A Flash v1.1]
#include <iostream>
#include <sstream>
#include <vector>
#include <fstream>

using namespace std;


class qSort {

  public:
    int pivot_count = 0;
    qSort(vector<int>& seq);

  private:
    void quick_sort(vector<int> seq, int begin, int end);
    void swap(int& a, int& b);

};


int main(int argc, char* argv[]) {
  ifstream inFile(argv[1]);
  vector<int> seq;
  seq.reserve(30); // Maximum number of elements
  int num;
  string line;

  while (getline(inFile, line)) {
    stringstream ss;
    ss.str(line);
    while (ss >> num) seq.push_back(num);
    cout << qSort(seq).pivot_count << "\n";
    seq.clear();
  }

  return(0);
}


qSort::qSort(vector<int>& seq) {
  int N = seq.size() - 1;
  quick_sort(seq, 0, N);
}


void qSort::quick_sort(vector<int> seq, int begin, int end) {
  if (begin < end) {
    pivot_count += 1;
    int pivot = seq[begin];
    int pivot_pos = begin;
    int left = begin+1;
    int right = end;

    while (left < right) {
      while (left < right && seq[right] >= pivot) right--;

      if (seq[right] < pivot) {
        swap(seq[pivot_pos], seq[right]);
        pivot_pos = right;
      }

      if (left < right) {
        while (left < right && seq[left] <= pivot) left++;

        if (seq[left] > pivot) {
          swap(seq[pivot_pos], seq[left]);
          pivot_pos = left;
        }
      }
    }

    quick_sort(seq, begin, pivot_pos-1);
    quick_sort(seq, pivot_pos+1, end);
  }
}

void qSort::swap(int& a, int& b) {
  int tmp = a;
  a = b;
  b = tmp;
}
